---
title: "tile a bed file"
output: html_document
date: "2024-12-31"
editor_options: 
  chunk_output_type: console
---

```{r}
library(GenomicRanges)
library(rtracklayer)
library(dplyr)
library(readr)
library(here)

```

```{r}
gr2 <- GRanges(
        seqnames=Rle(c("chr1", "chr2", "chr1", "chr3"), c(1, 3, 2, 4)),
        ranges=IRanges(1:10, end=11),
        strand=Rle(strand(c("-", "+", "*", "+", "-")), c(1, 2, 2, 3, 2)),
        seqlengths=c(chr1=11, chr2=12, chr3=13))

gr2 
# split every range to 2 base pair bins
tiles <- tile(gr2, width = 2L)
tiles
```


We want to tile it to 5 base pair bins
```{r}
gr<- rtracklayer::import(here("data/test.bed"))
gr

bin_size<- 5

gr_width<- width(gr)

gr_width

bin_num<- ceiling(gr_width/bin_size)

## after extending, the peaks are overlapping
gr_center<- resize(gr, fix = "center", width = bin_num * bin_size)

## this works fine
unlist(tile(gr_center, width = bin_size))

## merge before tile
gr_center_merge<- reduce(gr_center)

gr_center_merge

out_bed<- unlist(tile(gr_center_merge, width = bin_size))

out_bed
```

check `tileGenome`. It is very helpful when you need to bin the genome into
bins and calculate the number of reads in each bin.  For example, for scATACseq,
one can calculate the number of reads per bin in the whole genome and use that matrix
to cluster the cells.

```{r}

library(TxDb.Hsapiens.UCSC.hg38.knownGene)
txdb<- TxDb.Hsapiens.UCSC.hg38.knownGene

bins <- tileGenome(seqinfo(txdb), tilewidth=10000,
                   cut.last.tile.in.chrom=TRUE)
```

